osu.Game/Rulesets/Objects/SliderPathExtensions.cs

```diff
@@ -14,26 +17,28 @@ public static class SliderPathExtensions
         /// <summary>
         /// Snaps the provided <paramref name="hitObject"/>'s duration using the <paramref name="snapProvider"/>.
         /// </summary>
-        public static void SnapTo<THitObject>(this THitObject hitObject, IDistanceSnapProvider? snapProvider)
+        public static void SnapTo<THitObject>(this THitObject hitObject, IDistanceSnapProvider? snapProvider, NewBeatmapEditorChangeHandler? changeHandler = null)
             where THitObject : HitObject, IHasPath
         {
-            hitObject.Path.ExpectedDistance.Value = snapProvider?.FindSnappedDistance(hitObject, (float)hitObject.Path.CalculatedDistance, DistanceSnapTarget.Start) ?? hitObject.Path.CalculatedDistance;
+            double newDistance = snapProvider?.FindSnappedDistance(hitObject, (float)hitObject.Path.CalculatedDistance, DistanceSnapTarget.Start) ?? hitObject.Path.CalculatedDistance;
+            new ExpectedDistanceChange(hitObject.Path, newDistance).Apply(changeHandler);
         }
 
         /// <summary>
         /// Reverse the direction of this path.
         /// </summary>
         /// <param name="sliderPath">The <see cref="SliderPath"/>.</param>
         /// <param name="positionalOffset">The positional offset of the resulting path. It should be added to the start position of this path.</param>
-        public static void Reverse(this SliderPath sliderPath, out Vector2 positionalOffset)
+        /// <param name="changeHandler">Change handler to submit changes to.</param>
+        public static void Reverse(this SliderPath sliderPath, out Vector2 positionalOffset, NewBeatmapEditorChangeHandler? changeHandler = null)
         {
             var controlPoints = sliderPath.ControlPoints;
 
             var inheritedLinearPoints = controlPoints.Where(p => sliderPath.PointsInSegment(p)[0].Type == PathType.LINEAR && p.Type == null).ToList();
 
             // Inherited points after a linear point, as well as the first control point if it inherited,
             // should be treated as linear points, so their types are temporarily changed to linear.
-            inheritedLinearPoints.ForEach(p => p.Type = PathType.LINEAR);
+            inheritedLinearPoints.ForEach(p => new PathControlPointTypeChange(p, PathType.LINEAR).Apply(changeHandler));
```

Comment:
Q:
Should create a class inherit `CompositeChange` for list of changes?
I noticed that `CompositeChange` can do list of changes inside a change but have no idea when to use this.

Reply:
I forgot to remove the old `SliderPathExtensions`.

Every bit of edit code that you want to put into history should be an `IRevertibleChange`. If your change is not atomic then you use `CompositeChange`.
We could technically also make `CompositeChange` for other methods like `splitControlPoints`.

Reply:
It's a little bit hard to distinguish them.
e.g.
why add list of control points use the `AddRangePathControlPointChange`
but update list of control point use the `ForEach`

Also, guess should be OK to remove the `CompositeChange` if `InsertPathControlPointChange` can accept add list of control points?

e.g.
```csharp
public class InsertPathControlPointChange : IRevertibleChange
{
    public readonly IList<PathControlPoint> Target;

    public readonly int InsertionIndex;

    public readonly IList<PathControlPoint> Items;

    public InsertPathControlPointChange(IList<PathControlPoint> target, int insertionIndex, PathControlPoint item)
      : this(target, insertionIndex. new List<PathControlPoint> item)
    {
    }

    public InsertPathControlPointChange(IList<PathControlPoint> target, int insertionIndex, IList<PathControlPoint> items)
    {
        Target = target;
        InsertionIndex = insertionIndex;
        Items = items;
    }

    public void Apply() => Target.Insert(InsertionIndex, Items);

    public void Revert() => Target.RemoveAt(InsertionIndex, items.count);
}
```

Reply:
`AddRange` and `RemoveRange` are standard methods on any `IList`. It makes sense to have commands for these.

`ForEach` is a Linq extension method that takes a lambda for what to do. It's not very standard API and lambdas can't be serialized, so it doesn't make sense to make it a command.

I guess `AddRange` and `RemoveRange` could also be primitive changes (not `CompositeChange`) but for now I've left the set of primitive changes minimal.

Reply:
Ok i see.

Another question:
```
foreach (var h in hitObjects)
  new PositionChange(h, h.Position + localDelta).Apply(changeHandler);
```

seems can be:
```
new RelativePositionChange(hitObjects, localDelta).Apply(changeHandler);
```

Are you recommend to give the change a relative value(e.g. x += 3), or only accept the absolute value(x = 3).

Reply:
I prefer absolute changes over relative ones because they are more robust to changes in the history. If you have multiple move changes in history and remove one in the middle, then it wont matter for absolute changes, but with relative changes the result changes.

I'd need a good reason to add this beyond just saving one line of code.

Reply:
Agree with preferring absolute change commands over relative ones.